content/blog/2019-11-21-bug-finding/index.md

```diff
@@ -0,0 +1,308 @@
+
+There are several choices made because of the target:
+
+1. The Csmith designers manually tune the 80 probability variables to generate programs with a balanced mix of arithmetic and bitwise operations, of loops and straight-line code, and of single-level and multi-level indirections, *etc*.
+2. Encouraging Csmith to generate idiomatic code, e.g. loops that access all elements of an array.
+3. Discouraging Csmith from generating source-level diversity that is unlikely to improve the “coverage” of a compiler’s IR, e.g. additional levels of parentheses around expressions.
+4. Designing Csmith efficiently generates runnable programs of a few tens of thousands of lines long in a few seconds.
+
+## Evaluation
+
+### Opportunistic Bug Finding
+
+The Csmith designers tested 11 compilers and reports the errors to the developers. The commercial compiler developers do not care, while GCC and LLVM team response quickly. By the time the paper is formalized, 79 GCC bugs and 202 LLVM bugs (2% of all LLVM bug reports) are reported and most of them are fixed. CompCert is such a good compiler that the under-development version of CompCert is the only compiler Csmith cannot find wrong-code errors after 6 CPU-years of testing.
+
+### Bug Types
+
+Before we move on, I would like to introduce bug types for understanding the experiment results:
+
+1. A *crash error* is one that crashes the compiler during compilation.
+2. A *wrong error* is one that crashes the program during run-time.
+3. A *slient wrong error* is a wrong error that does not produce a compiler warning during compilation.
+4. An *assertion failure* is an LLVM internal consistency check failure.
+5. An *internal compiler failure* is a GCC internal consistency check failure.
+
+### Quantitative Comparison of GCC and LLVM Versions
+
+The following figure shows the compilation and execution results of LLVM 1.9–2.8, GCC 3.[0–4].0, and GCC 4.[0–5].0 given the input of 1,000,000 Csmith randomly generated programs. Every program was compiled at –O0, –O1, –O2, –Os, and –O3. A test case is valid if every compiler terminated within five minutes and if every compiled random program terminated within 5 seconds. The top and bottom row are the error rates of different compiler versions. The authors also find the source of the compilation error and plot it in the middle row.
+
+<img src="llvm_gcc.png" style="width: 100%">
+
+### Bug-Finding Performance as a Function of Test-Case Size
+
+The goal of designing Csmith is to find many defects quickly, and to what size the program should Csmith generate to achieve that goal becomes a question. When reporting the error, we preferred smaller programs over larger one because they are easier to debug and report. The figure below shows the experiment performed to learn the error number and runtime tradeoff given the same run-time. 
```

Comment:
"larger ones"

Also, is "we" the authors of the paper, or is it Chris and Zhijing?